### PR TITLE
Theta compression

### PR DESCRIPTION
**Pull request by AlexanderSaydakov**
compact compressed theta sketch
serialization version 4

### CODE REVIEW

**Comment by AlexanderSaydakov** [comment]
A couple of shortcuts for now:
1. wrap() delegates to heapify() for serial version 4. The alternative would be to have a special subclass such as DirectCompressedCompactSketch and a special support for it in set operations since they don't just iterate over any input sketch, but have special treatment (at least in the Union).
2. union(Memory) in case of serial version 4 calls heapify() and then delegates to union(Sketch). This can be implemented later as decompression during union to avoid full decompression in case of the early stop. This is a potential performance improvement.

**Comment by leerho** [comment]
I see quite a few CodeQL "notes" in the BitPacking class, all related to parameter "i".  Is this actually used somewhere making all of these notes false positives?

**Comment by AlexanderSaydakov** [comment]
That is not intended. My bad. Harmless though with the current usage. I will fix shortly.

**Comment by AlexanderSaydakov** [comment]
> @AlexanderSaydakov this looks really interesting - do you have any early results that compare the reduction introduced by the new serialisation?
> 
> Does it make sense to apply this to Tuple sketch encoding as well?

Here is the serialized size comparison.
Theta sketch compression

Of course, the compression affects the speed of serialization and deserialization. However, in a I/O bound system the overall performance might happen to be even better due to reduced size (I/O vs CPU tradeoff).
The compression is not done by default currently. The user has to explicitly call toByteArrayCompressed()

Regarding Tuple sketch. This method can be applied to the hash values in the Tuple sketch as well, but the benefit is not clear since the payload is likely to be much larger.

File: src/main/java/org/apache/datasketches/theta/CompactSketch.java [hunk 1]
```diff
@@ -285,4 +249,128 @@ public boolean isCompact() {
...
+    // assumes ordered sketch
+    long previous = 0;
+    long ored = 0;
+    HashIterator it = iterator();
+    while (it.next()) {
+      final long delta = it.get() - previous;
+      ored |= delta;
+      previous = it.get();
+    }
+    return Long.numberOfLeadingZeros(ored);
```
[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/theta/CompactSketch.java:270]: delta encoding makes the most sense for ordered values. this proposed code does not compress unordered sketches. toByteArrayCompressed() delegates to V3 toByteArray() for unordered, empty and single-item sketches.

File: src/main/java/org/apache/datasketches/theta/CompactSketch.java [hunk 2]
```diff
@@ -107,27 +91,34 @@ public static CompactSketch heapify(final Memory srcMem) {
...
     //not SerVer 3, assume compact stored form
-    final short seedHash = ThetaUtil.computeSeedHash(expectedSeed);
+    final short seedHash = ThetaUtil.computeSeedHash(seed);
     if (serVer == 1) {
       return ForwardCompatibility.heapify1to3(srcMem, seedHash);
     }
     if (serVer == 2) {
-      return ForwardCompatibility.heapify2to3(srcMem, seedHash);
+      return ForwardCompatibility.heapify2to3(srcMem,
+          enforceSeed ? seedHash : (short) extractSeedHash(srcMem));
```
[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/theta/CompactSketch.java:121]: I think that those methods names heapify1to3 and heapify2to3 are a bit confusing. The serialization format does not really apply to on-heap sketch. And now with V4 introduced they are even more confusing. Perhaps we should rename them.
The proposed code does not force ordering. Instead it serializes unordered sketches using the current V3 format. This is debatable. It was proposed that toByteArrayCompressed() should force ordering, but we did not reach that agreement yet. How would you suggest handling this?

[the remark this answers is not in the document]
Comment (davecromberge) [reply src/main/java/org/apache/datasketches/theta/CompactSketch.java:121]: I think leaving the decision up to the library user is sensible. As you mentioned in another comment, am I/O bound workload may require different tradeoffs to a CPU bound workload.

As an aside, I downloaded your branch locally to experiment with the serialisation and as a user I was expecting the compressed serialisation to apply automatically. I think the release notes and documentation could probably draw attention to this fact.

**Comment by davecromberge** [comment]
That is a nice improvement! Thanks for sharing the characterisation result. The reasoning about Tuple makes sense as well.

**Review by leerho — APPROVED** [review]
Your last commit fixed the CodeQL notes issue, so I am happy.

File: src/main/java/org/apache/datasketches/theta/CompactSketch.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/theta/CompactSketch.java:270]: deltas are not ordered unlike the underlying hashes, so the deltas of deltas can be negative, which messes up the approach based on leading zeros. I am not sure how to compress better and still be fast. Ideas are welcome.

**Pull request merged**